### ISSUE TITLE
Chapter 5: centering the training data for LinearSVC in a sparse matrix

### ISSUE THREAD

**Issue by genemishchenko** [ageron/handson-ml#481]
Hi Aurelien.

In Chapter 5 on SVMs you write:

> The LinearSVC class regularizes the bias term, so you should center the training set first by subtracting its mean. This is automatic if you scale the data using the StandardScaler.

The StandardScaler, however, does NOT work on sparse matrices, which are very common in NLP applications (CountVectorizer's and TfidfTransformer's output type is csr_matrix, for instance).
This may be worth noting in the book. 

That was the comment part... I also have a question based on it:

I found this great sklearn documentation page describing in detail what all the scalers do:
[https://scikit-learn.org/stable/auto_examples/preprocessing/plot_all_scaling.html](https://scikit-learn.org/stable/auto_examples/preprocessing/plot_all_scaling.html)
(it focuses on managing outliers, but it's a great general visual overview)

It does not look like there is any other scaler that can do the same thing as StandardScaler does with dense arrays, but also works on sparse matrices, so I have implemented a custom transformer that does this:
```
global_mean = X_train_tfidf.sum() / X_train_tfidf.nnz
X_train_tfidf.data -= global_mean
```
Is this correct and sufficient or can you recommend a better method?

Thank you.
Gene.

**Comment by genemishchenko** [comment]
Actually, the biggest issue that I have is that centering the data absolutely kills the accuracy of LinearSVC.
After applying the self-made centering method above on the training sparse matrix of 130K features and 12K instances with 1.4M total data points LinearSVC gives me only 90% accuracy on the full training set and the measly 50% on the test set.
Without centering or scaling of any kind I get 99.5% and 85% correspondingly (I tried both SGDClassifier and LinearSVC, resulting in the same maximum accuracy for the data at hand).
Am I doing something wrong or does centering the data actually hurt the LinearSVC performance?
